### ISSUE TITLE
How to manually set the seed of randomsampler generator when training using transformers trainer

### ISSUE THREAD

**Issue by young-chao** [huggingface/transformers#26218]
### System Info

I used a [script](https://github.com/huggingface/transformers/blob/v4.33.0/examples/pytorch/language-modeling/run_clm.py) to continue pre-training the llama2 model. In the second epoch, the loss began to explode, so I chose to reload the checkpoint to continue training, but the loss changes were completely consistent with before, which made me doubt the iteration of the dataset is always consistent. So I tried modifying the [seed.](https://github.com/huggingface/transformers/blob/v4.33.0/examples/pytorch/language-modeling/run_clm.py#L309C33-L309C33) But in the end, my training loss is always consistent, and the state I print randomsampler is always the same. 
I hope someone can tell me how to solve this problem, including where the seed of this generator is specified.

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

transformers==4.33.0
pytorch==1.13.1
accelerate==0.21.0
deepspeed==0.10.0

### Expected behavior

I hope that the sampling of training data set should be different every time.

**Comment by young-chao** [comment]
![image](https://github.com/huggingface/transformers/assets/34190033/af693746-4e27-4c6b-bf5b-0b58ef173a38)
I found that the default initialized state of the generator is always this tensor, regardless of the cuda version, pytorch version, etc.

**Comment by ArthurZucker** [comment]
Two things you can try: 
- pass a generator 
- use `set_seed(X)` which is a utility trainer.

**Comment by young-chao** [comment]
你好，你的邮件已收到，谢谢！

**Comment by young-chao** [comment]
> Two things you can try:
> 
> * pass a generator
> * use `set_seed(X)` which is a utility trainer.

Thank you for your reply, I will use your solution.

**Comment by young-chao** [comment]
@ArthurZucker  I am going to try your solution, but I found that the trainer does not receive dataloader or sampler as a parameter, so where should I pass the generator to the sampler. Moreover, the trainer does not have this set_seed method. Do you mean the set_seed method of transformers? The script I use uses this method, however it only affects the default generator.

**Comment by young-chao** [comment]
你好，你的邮件已收到，谢谢！
